### PR TITLE
revert Improve effect analysis of bitshifts #47567

### PR DESCRIPTION
**Pull request by oscardssmith**
This broke the BitIntegers package. It would be good to fix the original issue from the compiler side.

### CODE REVIEW

**Review by quinnj — APPROVED** [review]

**Comment by giordano** [comment]
> It would be good to fix the original issue from the compiler side.

Open an issue to keep track of that?

**Comment by KristofferC** [comment]
> This broke the BitIntegers package.

Okay but why? What is the bug in this PR and should a corresponding test be added so that it doesn't regress again?

File: test/int.jl [hunk 1]
```diff
@@ -363,7 +369,7 @@ end
                 @test div(x,-y,rnd) == round(x/-y,rnd)
             end
             @test divrem(x,y,RoundFromZero) == (div(x,y,RoundFromZero), rem(x,y,RoundFromZero))
-            @test divrem(x,-y,RoundFromZero) == (div(x,-y,RoundFromZero), rem(x,-y,RoundFromZero))
+            @test divrem(x,-y,RoundFromZero) == (div(x,-y,RoundFromZero)z, rem(x,-y,RoundFromZero))
```
Comment (KristofferC) [inline test/int.jl:372]: Hm?

**Comment by oscardssmith** [comment]
@KristofferC it wasn't a bug, just that the way it widened applicable types caused ambiguity errors for packages that tried to extend it.

**Comment by aviatesk** [comment]
> just that the way it widened applicable types caused ambiguity errors for packages that tried to extend it.

Could you add a simplified test case, i.e. the operation that caused the issue now doesn't widen them?

**Comment by aviatesk** [comment]
@oscardssmith bump?

**Comment by oscardssmith** [comment]
sorry, I did a bad job explaining what is wrong here. the PR didn't change any values, but meant that new types wouldn't work. I'm not sure there is a great way to test for this.

File: test/int.jl [hunk 2]
```diff
@@ -201,7 +201,13 @@ end
         end
         for T2 in Base.BitInteger_types
             for op in (>>, <<, >>>)
-                @test Core.Compiler.is_total(Base.infer_effects(op, (T, T2)))
+                if sizeof(T2)==sizeof(Int) || T <: Signed || (op==>>>) || T2 <: Unsigned
+                    @test Core.Compiler.is_total(Base.infer_effects(op, (T, T2)))
+                else
+                    @test Core.Compiler.is_foldable(Base.infer_effects(op, (T, T2)))
+                    # nothrow analysis of recursion needs fixing
+                    @test_broken Core.Compiler.is_total(Base.infer_effects(op, (T, T2)))
```
Comment (aviatesk) [inline test/int.jl:209]: Don't use `:total` but use as accurate effect as possible
```suggestion
                    # #47835, TODO implement interval arithmetic analysis
                    @test_broken Core.Compiler.is_nothrow(Base.infer_effects(op, (T, T2)))
```

**Pull request merged**